src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -825,27 +824,91 @@ public void BlockDataDownloadedIsCalled_ForValidBlocksAfterFv_ResultShouldBeTrue
-            bool resultForH6 = cht.BlockDataDownloaded(chainTipH6, ctx.CreateBlock());
-            bool resultForH5 = cht.BlockDataDownloaded(chainTipH5, ctx.CreateBlock());
+            // Call BlockDataDownloaded on h1, h2 and h3.
+            ChainedHeader chainTipH3 = chainTip;
+            ChainedHeader chainTipH2 = chainTip.Previous;
+            ChainedHeader chainTipH1 = chainTipH2.Previous;
+            bool resultForH3 = cht.BlockDataDownloaded(chainTipH3, ctx.CreateBlock());
+            bool resultForH2 = cht.BlockDataDownloaded(chainTipH2, ctx.CreateBlock());
+            bool resultForH1 = cht.BlockDataDownloaded(chainTipH1, ctx.CreateBlock());
 
-            // Blocks should be set and only header 5 result is true.
-            resultForH7.Should().BeFalse();
-            chainTipH7.Block.Should().NotBeNull();
+            // Blocks should be set and only header 1 result is true.
+            resultForH3.Should().BeFalse();
+            chainTipH3.Block.Should().NotBeNull();
 
-            resultForH6.Should().BeFalse();
-            chainTipH6.Block.Should().NotBeNull();
+            resultForH2.Should().BeFalse();
+            chainTipH2.Block.Should().NotBeNull();
 
-            resultForH5.Should().BeTrue();
-            chainTipH5.Block.Should().NotBeNull();
+            resultForH1.Should().BeTrue();
+            chainTipH1.Block.Should().NotBeNull();
+        }
+
+        /// <summary>
+        /// Issue 25 @ Consensus and headers are at block 5. Block 6 is presented and PV is successful.
+        /// Call PartialOrFullValidationFailed and after make sure that there is a single US_CONSTANT on CT.
+        /// </summary>
+        [Fact]
+        public void ConsensusAndHeadersAreAtBlock5_Block6Presented_PartialOrFullValidationFailedCalled_ThereShouldBeASingleConstant()
+        {
+            // Chain header tree setup. Initial chain has 5 headers.
+            // Example: h1=h2=h3=h4=h5.
+            const int initialChainSize = 5;
+            TestContext ctx = new TestContextBuilder()
+                .WithInitialChain(initialChainSize)
+                .Build();
+            ChainedHeaderTree cht = ctx.ChainedHeaderTree;
+            ChainedHeader chainTip = ctx.InitialChainTip;
+
+            // Extend a chain by 1 header.
+            // Example: h1=h2=h3=h4=h5=h6.
+            chainTip = ctx.ExtendAChain(1, chainTip);
+            List<BlockHeader> listOfCurrentChainHeaders = ctx.ChainedHeaderToList(chainTip, 1);
+
+            // Present header by peer with id 1 and then call PartialOrFullValidationFailed on it.
+            const int peerId = 1;
+            cht.ConnectNewHeaders(peerId, listOfCurrentChainHeaders);
+            List<int> validationResult = cht.PartialOrFullValidationFailed(chainTip);
+
+            // Validation result should contain 1 result pointing to CT.
+            validationResult.Should().HaveCount(1);
+            validationResult.First().Should().Be(peerId);
+        }
```

Comment:
In this test, we first need to call partial validation succeeded and then we want to call partial of validation failed on the same block header. Then we need to check the internal structures looking for the local PID. It must be found only once on header five and must not be present on six.

Reply:
see peerIdsByTipHash

Reply:
This is not what test description was saying. Ok I will change it.

Reply:
The descriptions are sometimes cryptic– This is why I said that the descriptions put into the summary should not be just a copy and paste, because it might not make good sense. The description is not clear, it is better to ask for explanation before starting implementation. Here the description says that PV was successful, meaning that partial validation succeeded was supposed to be called. The us constant in the description refers to the local PID.

Reply:
So maybe I get it wrong, let's wait for @noescape00 two confirm the correct approach.

Reply:
Ok, the logic is a bit incomplete. 

First we should present the headers (ConnectNewHeaders), after that mark them as PV (call PartialValidationSucceeded) and after that fail full validation (call PartialOrFullValidationFailed) and ensure that there is only one header claimed by LocalPeerId and it is equal to old tip before reorg was tried.

Reply:
So logically 2 things are missing in this test- marking the header as PV and checking the inner structures to ensure that LocalPeerId is where it should be.